### DISCUSSION TITLE
How do I prevent `output: asis` chunks from modifying `htmltools` outputs?

### DISCUSSION THREAD

**Discussion by AlbertRapp** [quarto-dev/quarto-cli#7614] [Q&A] [answered]
### Description

This is related to #7600. I'm pretty confused why `output: asis` modifies HTML code. In this reprex, the rendered HTML-doc has additional `p`-tags even though they are not in the output that is generated from the R code chunk. Is there a way to circumvent that to get the generated HTML code from the R chunk "as is"?

````qmd
```{r}
#| echo: false
#| output: asis

htmltools::div(
  htmltools::a('Some link')
) |> as.character() |> cat(sep = "\n")
```
````

The output from this chunk (if executed in R in the console) is

```
<div>
  <a>Some link</a>
</div>
```

But once I render in Quarto I get

```
<div>
  <p>
    <a href>Some link</a>
  </p>
</div>
```

**Comment by cscheid** [comment]
Can you provide a full example of your input to quarto? Thank you.

**Reply by AlbertRapp** [reply]
The R code chunk that I provided is the only thing in my Quarto doc. And the two chunks that I posted after the R code chunk just describe what I see in the R console vs the HTML output.

**Reply by mcanouil** [reply]
You don't have frontmatter specifying at least the format?
Below is a complete small reproducible example that anyone can simply copy, paste, and run without having to "guess" anything.

`````md
````qmd
---
title: "Reproducible Quarto Document"
format: html
---

This is a reproducible Quarto document using `format: html`.
It is written in Markdown and contains embedded R code.
When you run the code, it will produce a plot.

```{r}
plot(cars)
```

The end.
````
`````

**Reply by AlbertRapp** [reply]
Nope. No YAML header. Trying to keep the example as minimal as possible by taking a blank doc and inserting that one code chunk. But I can add a header next time and share it.

**Reply by mcanouil** [reply]
You still render to one format 😅
Adding the format in the frontmatter ensure we are on the same page.

**Comment by cscheid** [comment, marked as answer]
What's happening here is that Pandoc doesn't like `<a>` elements to be directly inside `divs`, and so it inserts a paragraph. This is happening because your output (use `keep-md: true` to inspect it) doesn't bracket the content in a raw block:

```
---
keep-md: true
---


<div>
  <a>Some link</a>
</div>
```

If you need your HTML output to be unprocessed by Pandoc, you need to emit a raw block. I don't know the htmltools syntax to do it, but `knitr::raw_html` will take a string as input and produce a raw block.

````
---
keep-md: true
---

```{r}
#| echo: false
#| output: asis

knitr::raw_html(
htmltools::div(
  htmltools::a('Some link')
) |> as.character() |> paste())
```
````

produces

````
---
keep-md: true
---


::: {.cell-output-display}


```{=html}
<div>
  <a>Some link</a>
</div>
```


:::
````

which then is unprocessed by Pandoc.

**Reply by cscheid** [reply]
(nevermind the `cell-output-display` business, _that's_ for the other issue you've raised)

**Reply by AlbertRapp** [reply]
Ah perfect. That's the answer I was looking for. FYI it also works to replace `paste()` by `raw_html()`, i.e.

````qmd

```{r}
#| echo: false
#| output: asis


htmltools::div(
  htmltools::a('Some link'),
) |> as.character() |> knitr::raw_html()
```

````

**Reply by cderv** [reply]
See below how I would recommend to do it (https://github.com/quarto-dev/quarto-cli/discussions/7614#discussioncomment-7597596)

**Comment by cderv** [comment]
This should be as simple as that to use **knitr** and **htmltools** to produce HTML content 

````markdown
---
title: "test"
format: html
keep-md: true
---

```{r}
#| echo: false
#| output: asis
htmltools::div(
  htmltools::a('Some link')
)
```

````

![image](https://github.com/quarto-dev/quarto-cli/assets/6791940/0aa81528-cc22-494d-9502-05b0ab2444e6)

**htmltools** + **knitr** takes care for you about all the thing you need to do to print the right markdown. 

````markdown
---
title: "test"
format: html
keep-md: true
---


::: {.cell-output-display}


```{=html}
<div>
<a>Some link</a>
</div>
```


:::

````

(and as said, the `.cell-output-display` will be removed when fixing #7600). 

So no need of `as.character()` or `raw_block()` because **knitr** and **htmltools** does it for you (through the clever [custom print](https://cran.r-project.org/web/packages/knitr/vignettes/knit_print.html) logic from knitr. 

That is to me the right way to do that. 

Does it work for you correctly ? Did you try it ? If not working, which version of the tools are you using ?

**Reply by AlbertRapp** [reply]
Yes, that's how I feel too. And it works just like you said. But it doesn't work with the `|> as.character() |> cat(sep = "\n")` approach. And I used that approach because I had to use that way in #7600 and so I thought this is the safer version. Turns out that was wrong :laughing: Good thing that the `.cell-output-display` thing will be taken care of. Thank you for all the help :partying_face:

**Reply by cderv** [reply]
> But it doesn't work with the |> as.character() |> cat(sep = "\n") approach.

it does not do the same thing. Some information are lost in translation when you do that, like the fact a raw HTML block is required - hence why you need to add it yourself using `raw_block()` if you go this road.
